### PR TITLE
Backend user authentication

### PR DESCRIPTION
**Pull request by mpodgorny**
When logging in, server saves two cookies - refresh_token (long expiration date) and acces _token (short expiration date). On every action server will validate acces_token (for future implementation); when acces_token expires server will send HTTP 401 error, on which client should automatically go to GET /get-acces-token. This endpoint will verify refresh_token and set new acces_token if succeed, then client should redirect again to previously requested resource.

POST /register  takes email, username, password, rePassword in body and verifies if username and/or email is taken - adequate response code is sent. Next step will require connecting to database to save changes.
Password are hashed using HS256 and will be stored in db.

### CODE REVIEW

File: backend/src/routers/user_auth/get_token.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        )
+    username = verify_refresh_token(refresh_token)
+    if username:
+        access_token_expires = timedelta(minutes=ACCES_TOKEN_EXPIRES_MINUTES)
+        acces_token = create_token(
+            data={"sub": username}, expires_delta=access_token_expires
+        )
+        acc_token = jsonable_encoder(acces_token)
+        response.set_cookie(
+            key="access_token", value=f"{acc_token}", httponly=True,
```
Comment (wadamek65) [inline backend/src/routers/user_auth/get_token.py]: `value=f"{acc_token}"` this looks weird. Did you forget something or do you want to cast the `acc_token` to a string? If you wanted to do the latter `str(acc_token)` should work fine.

Comment (wadamek65) [reply backend/src/routers/user_auth/get_token.py]: Also, I think that you want to set the expiration time of the cookie

File: backend/src/routers/user_auth/login.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    if not verify_password(password, user.hashed_password):
+        return False
+    return user
+
+
+@router.post("/login")
+async def login_for_access_token(form_data: OAuth2PasswordRequestForm = Depends()):
+    user = authenticate_user(MOCK_DB, form_data.username, form_data.password)
+    if not user:
+        print("User NOT verified.")
```
Comment (wadamek65) [inline backend/src/routers/user_auth/login.py]: Instead of `print`s, consider using `logging` package so that you can control the logging output between production and development (I know it is really not necessary for you since this isn't a very serious project, but it's good to know for the future). Also add a little more information to the log so it is more useful, like the username or the reason why it failed.

File: backend/src/routers/user_auth/register.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+from starlette.status import HTTP_417_EXPECTATION_FAILED, HTTP_409_CONFLICT
+from starlette.responses import Response
+from pydantic import BaseModel
+from .utils import MOCK_DB, get_password_hash, User
+
+
+router = APIRouter()
+
+
+class Register_form(BaseModel):
```
Comment (wadamek65) [inline backend/src/routers/user_auth/register.py]: Since this is a class, the name should be in pascal case.

File: backend/src/routers/user_auth/register.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    username: str
+    email: str
+    password: str
+    rePassword: str
```
Comment (wadamek65) [inline backend/src/routers/user_auth/register.py]: This is a class field and should be in snake case.

File: backend/src/routers/user_auth/register.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    # There will be adding to database, but db is required.
+    print("User: ", user["username"], "added.")
+
+
+@router.post("/register")
+async def register(form_data: Register_form):
+    if form_data.password != form_data.rePassword:
+        print("Passwords were not the same.")
+        raise HTTPException(
+            status_code=HTTP_417_EXPECTATION_FAILED,
```
Comment (wadamek65) [inline backend/src/routers/user_auth/register.py]: This response should be used only when requests provide the `Expect` header, see here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expect

A simple `400 Bad Request` looks better here.

File: backend/src/routers/user_auth/register.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    user = {
+        "username": form_data.username,
+        "email": form_data.email,
+        "hashed_password": hashed_password,
+        "disabled": False,
+    }
+    if not check_user_uniqueness(MOCK_DB, user):
+        print("User already taken.")
+        raise HTTPException(
+            status_code=HTTP_409_CONFLICT, detail="Username or email already taken.",
```
Comment (wadamek65) [inline backend/src/routers/user_auth/register.py]: This status code is also not an ideal use case. A simple bad request works better here too. You would want to use `409` in case of a conflict that you'd need to resolve in some other way and that would allow you to send the same request and make it pass with 200. This is not the case here, once a username is taken the user has no means of conflict resolution, he will have to change the username and retry the request with different parameters.

File: frontend/Pipfile [hunk 7]
```diff
@@ -0,0 +1,11 @@
+[[source]]
```
Comment (wadamek65) [inline frontend/Pipfile]: I don't think you want to add this file.

File: frontend/package.json [hunk 8]
```diff
@@ -6,9 +6,11 @@
     "@material-ui/core": "^4.5.2",
     "@material/card": "^4.0.0",
     "bootstrap": "^4.3.1",
+
```
Comment (wadamek65) [inline frontend/package.json]: Unnecessary whitespace

File: frontend/src/components/Auth/Login/Login.js [hunk 9]
```diff
@@ -6,22 +6,41 @@ import { Formik, Form } from "formik";
...
-const form = {
-  email: "",
-  password: ""
-};
+async function handleSubmit(params) {
+  const FD = new FormData();
+  FD.append("username", params.data.username);
+  FD.append("password", params.data.password);
+
+  fetch("http://127.0.0.1:8000/login", {
```
Comment (wadamek65) [inline frontend/src/components/Auth/Login/Login.js]: Just a reminder: `ky` will probably be more convenient than `fetch` https://github.com/sindresorhus/ky

File: frontend/src/components/Auth/Login/Login.js [hunk 10]
```diff
@@ -6,22 +6,41 @@ import { Formik, Form } from "formik";
...
+    method: "POST",
+    body: FD
+  })
+    .then(response => {
+      if (response.ok) {
+        return response.json();
+      }
+      return Promise.reject(new Error("Something is not yes."));
```
Comment (wadamek65) [inline frontend/src/components/Auth/Login/Login.js]: This error message sounds just perfect.

**Review by wadamek65 — APPROVED** [review]

File: backend/src/routers/user_auth/get_token.py (hunk 1 above)
Comment (mpodgorny) [reply backend/src/routers/user_auth/get_token.py]: You're right, also according to [jsonable_encoder docs](https://fastapi.tiangolo.com/tutorial/encoder/) i can just do `value=acc_token`. Change coming in next commit.

File: backend/src/routers/user_auth/login.py (hunk 2 above)
Comment (mpodgorny) [reply backend/src/routers/user_auth/login.py]: Actually, uvicorn gives nicely described logs by itself (Im sure they come from package `logging` you mentioned); I'll keep it in mind and try to improve logging in future.

File: backend/src/routers/user_auth/register.py (hunk 6 above)
Comment (mpodgorny) [reply backend/src/routers/user_auth/register.py]: Changed it, thanks for use case example, I'll have to do some more digging on http protocols.

File: frontend/src/components/Auth/Login/Login.js (hunk 10 above)
Comment (mpodgorny) [reply frontend/src/components/Auth/Login/Login.js]: Ponglish ftw ( ͡° ͜ʖ ͡°)

File: frontend/src/components/Auth/Login/Login.js (hunk 9 above)
Comment (mpodgorny) [reply frontend/src/components/Auth/Login/Login.js]: Thanks! I'll keep it as example to rework using `ky` in near future, when this part of app will be build.

File: backend/src/routers/user_auth/register.py (hunk 6 above)
Comment (wadamek65) [reply backend/src/routers/user_auth/register.py]: Most of the status codes are used in very specific use cases. Most of the time 400 to 404 codes are the best to use, but can't go wrong with any of them as long as the error is clear and descriptive.

File: backend/src/routers/user_auth/get_token.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+@router.get("/get-acces-token")
+async def get_acces_token(request: Request):
+    refresh_token = request.cookies.get("refresh_token")
+    response = Response()
+    if refresh_token is None:
+        raise HTTPException(
+            status_code=HTTP_403_FORBIDDEN, detail="No rights to refresh token.",
+        )
+    username = verify_refresh_token(refresh_token)
+    if username:
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/get_token.py]: Consider explicit comparison to `None`:
```python
if username is not None:
    ...
```

File: backend/src/routers/user_auth/utils.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    to_encode.update({"exp": expire})
+    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
+    return encoded_jwt
+
+
+def verify_refresh_token(token):
+    payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
+    username: str = payload.get("sub")
+    if not get_user(MOCK_DB, username):
+        raise HTTPException(
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/utils.py]: Where do you handle this `HTTPException`?

File: backend/src/routers/user_auth/get_token.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    verify_refresh_token,
+    ACCES_TOKEN_EXPIRES_MINUTES,
+    create_token,
+)
+
+
+router = APIRouter()
+
+
+@router.get("/get-acces-token")
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/get_token.py]: typo

File: backend/Pipfile [hunk 14]
```diff
@@ -15,8 +15,15 @@ pre-commit = "*"
 fastapi = "*"
 uvicorn = "*"
 email-validator = "*"
+python-multipart = "*"
```
Comment (pawelrubin) [inline backend/Pipfile]: Are you sure you need this package?

File: backend/src/main.py [hunk 15]
```diff
@@ -1,8 +1,28 @@
...
+# Development puprose only
+origins = [
+    "http:localhost",
+    "http:localhost:8080",
+]
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=origins,
+    allow_credentials=True,
+    allow_methods=["*"],
```
Comment (pawelrubin) [inline backend/src/main.py]: Are you sure you want to allow all methods, including `DELETE`?

File: backend/src/routers/user_auth/get_token.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    refresh_token = request.cookies.get("refresh_token")
+    response = Response()
+    if refresh_token is None:
+        raise HTTPException(
+            status_code=HTTP_403_FORBIDDEN, detail="No rights to refresh token.",
+        )
+    username = verify_refresh_token(refresh_token)
+    if username:
+        access_token_expires = timedelta(minutes=ACCES_TOKEN_EXPIRES_MINUTES)
+        acces_token = create_token(
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/get_token.py]: typo

File: backend/src/routers/user_auth/get_token.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+            data={"sub": username}, expires_delta=access_token_expires
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/get_token.py]: `sub`? If you meant `subject`, please be more descriptive.

File: backend/src/routers/user_auth/get_token.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+        )
+        acc_token = jsonable_encoder(acces_token)
+        response.set_cookie(
+            key="access_token", value=f"{acc_token}", httponly=True, expires=1000,
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/get_token.py]: According to [`encoder` docs](https://fastapi.tiangolo.com/tutorial/encoder/), you don't need to parse already parsed token to a `str`:
```python
response.set_cookie(
            key="access_token", value=jsonable_encoder(acces_token), httponly=True, expires=1000,
```

File: backend/src/routers/user_auth/login.py [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+router = APIRouter()
+
+
+def authenticate_user(fake_db, username: str, password: str):
+    user = get_user(fake_db, username)
+    if not user:
+        return False
+    if not verify_password(password, user.hashed_password):
+        return False
+    return user
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/login.py]: If `authenticate_user` returns `User` instance, you should return `None` in case of failure.
Moreover consider explicit comparison to `None`.
```Python
    user = get_user(fake_db, username)
    if user is None or not verify_password(password, user.hashed_password):
        return None
    return user
```

File: backend/src/routers/user_auth/login.py [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+
+@router.post("/login")
+async def login_for_access_token(form_data: OAuth2PasswordRequestForm = Depends()):
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/login.py]: I think `login` is enough
```python
async def login(...)
```

File: backend/src/routers/user_auth/utils.py [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+
+
+class User(BaseModel):  # pylint: disable=R0903
+    username: str
+    email: str = None
+    full_name: str = None
+    disabled: bool = None
+
+
+class UserInDB(User):  # pylint: disable=R0903
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/utils.py]: Why two classes for a user?

File: frontend/src/components/Auth/Register/Register.js [hunk 22]
```diff
@@ -6,13 +6,46 @@ import { Formik, Form } from "formik";
...
+        case 409:
+          window.alert("Username or email already taken.");
+          break;
+        case 417:
+          window.alert(
+            "You've manage to modify formik verification, you dirty hackerman!"
+          );
+          break;
+        default:
+          window.alert("Some other error, nice");
```
Comment (pawelrubin) [inline frontend/src/components/Auth/Register/Register.js]: Hold on buddy, we don't do `window.alert`s here XD

File: backend/src/routers/user_auth/utils.py [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def create_token(*, data: dict, expires_delta: timedelta = None):
+    to_encode = data.copy()
+    if expires_delta:
+        expire = datetime.utcnow() + expires_delta
+    else:
+        expire = datetime.utcnow() + timedelta(minutes=15)
+    to_encode.update({"exp": expire})
+    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/utils.py]: `ALGORITHM` does not tell anything about actual implementation.

File: backend/src/routers/user_auth/utils.py [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+def get_user(db, username: str):
+    if username in db:
+        user_dict = db[username]
+        return UserInDB(**user_dict)
+    return None
+
+
+def create_token(*, data: dict, expires_delta: timedelta = None):
+    to_encode = data.copy()
+    if expires_delta:
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/utils.py]: Again, if you are checking for `None`, please be explicit

File: backend/src/routers/user_auth/utils.py [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+def verify_password(plain_password, hashed_password):
+    return PWD_CONTEXT.verify(plain_password, hashed_password)
+
+
+def get_password_hash(password):
+    return PWD_CONTEXT.hash(password)
+
+
+def get_user(db, username: str):
+    if username in db:
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/utils.py]: Hmmm, `db` is a dict in a form of:
```python
{
    "foo": {
        "username": "bar",
        ...
    }
}
```
Let `username = "bar"`, 
```python 
if username in db:
    print("xD")
````
This code will not print `xD`.
I'm not sure if it's a behavior you wanted to implement.

File: backend/src/routers/user_auth/utils.py [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+class UserInDB(User):  # pylint: disable=R0903
+    hashed_password: str
+
+
+def verify_password(plain_password, hashed_password):
+    return PWD_CONTEXT.verify(plain_password, hashed_password)
+
+
+def get_password_hash(password):
+    return PWD_CONTEXT.hash(password)
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/utils.py]: So we'll store simple hashes only? No *salt* or *repetitions*?
Consider using [pbkdf2_sha](https://passlib.readthedocs.io/en/stable/lib/passlib.hash.pbkdf2_digest.html)

File: backend/src/routers/user_auth/utils.py (hunk 25 above)
Comment (mpodgorny) [reply backend/src/routers/user_auth/utils.py]: utils.py now have real database functionality, take a look at latest commit.

File: backend/Pipfile [hunk 27]
```diff
@@ -10,20 +10,22 @@ flake8 = "*"
...
 bcrypt = "*"
 datetime = "*"
-starlette = "*"
 pyjwt = "*"
 mongoengine = "*"
 atomicwrites = "*"
 pydantic = "*"
+starlette = "*"
+python-multipart = "*"
+pylint-mongoengine = "*"
```
Comment (pawelrubin) [inline backend/Pipfile]: Please reinstall this package with `--dev` since it's a dev one.

File: backend/src/routers/user_auth/get_token.py [hunk 28]
```diff
@@ -23,14 +23,18 @@
...
+            status_code=HTTP_403_FORBIDDEN, detail="Username not found.",
         )
 
+    access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRES_MINUTES)
+    acces_token = create_token(
+        data={"sub": username}, expires_delta=access_token_expires
+    )
+    acc_token = jsonable_encoder(acces_token)
+    response.set_cookie(
+        key="access_token", value=acc_token, httponly=True, expires=1000,
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/get_token.py]: 1. typo xD
2. suggestion:
```python
key="access_token", value=jsonable_encoder(access_token), httponly=True, expires=1000,
```

File: backend/src/routers/user_auth/login.py [hunk 29]
```diff
@@ -3,59 +3,60 @@
...
-    user = authenticate_user(MOCK_DB, form_data.username, form_data.password)
-    if not user:
+async def login(form_data: OAuth2PasswordRequestForm = Depends()):
+    user = authenticate_user(form_data.username, form_data.password)
+    if user is None:
         raise HTTPException(
-            status_code=HTTP_401_UNAUTHORIZED,
-            detail="Incorrect username or password",
-            headers={"WWW-Authenticate": "Bearer"},
+            status_code=HTTP_400_BAD_REQUEST, detail="Incorrect username or password",
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/login.py]: Shouldn't it be `HTTP_401_UNAUTHORIZED`?

File: backend/src/routers/user_auth/login.py [hunk 30]
```diff
@@ -3,59 +3,60 @@
...
+        value=jsonable_encoder(refresh_token),
+        httponly=True,
+        expires=36000,
     )
     response.set_cookie(
-        key="access_token", value=acc_token, httponly=True, expires=1000,
+        key="access_token",
+        value=jsonable_encoder(access_token),
+        httponly=True,
+        expires=1000,
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/login.py]: magic number here,
please change to `expires=<a variable>`,

File: backend/src/routers/user_auth/register.py [hunk 31]
```diff
@@ -1,56 +1,60 @@
...
-        ):
-            return False
-    return True
+class RegisterForm:  # pylint: disable=R0903
+    def __init__(
+        self,  # pylint: disable=C0330
+        username: str = Form(...),  # pylint: disable=C0330
+        email: str = Form(...),  # pylint: disable=C0330
+        password: str = Form(...),  # pylint: disable=C0330
+        re_password: str = Form(...),  # pylint: disable=C0330
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/register.py]: The default values are `Form(...)`, why exactly?

File: backend/src/routers/user_auth/register.py [hunk 32]
```diff
@@ -1,56 +1,60 @@
...
-    # There will be adding to database, but db is required.
-    logging.info("User: %s  added to DB (curretnly mocked)", user["username"])
+    try:
+        user.save()
+    except SaveConditionError:
+        return None
+    except NotUniqueError:
+        return None
+    logging.info("User: %s  added to DB", user["username"])
+    return user
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/register.py]:
```python
    try:
        user.save()
        logging.info("User: %s  added to DB", user["username"])
        return user
    except (SaveConditionError, NotUniqueError):
        return None
   
```

File: backend/src/routers/user_auth/register.py [hunk 33]
```diff
@@ -1,56 +1,60 @@
...
+        disabled=False,
+    )
+    if add_user_to_db(user) is None:
         logging.info("User: %s already taken", user["username"])
         raise HTTPException(
             status_code=HTTP_400_BAD_REQUEST, detail="Username or email already taken.",
         )
-    add_user_to_db(user)
     response = Response(status_code=200)
     return response
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/register.py]:
```python
return Response(status_code=200)
```

File: backend/src/routers/user_auth/utils.py [hunk 34]
```diff
@@ -1,46 +1,51 @@
...
-class User(BaseModel):  # pylint: disable=R0903
-    username: str
-    email: str = None
-    full_name: str = None
-    disabled: bool = None
+class User(Document):
+    username = StringField(required=True, max_length=200)
+    email = StringField(required=True, max_length=200)
+    hashed_password = StringField(required=True, max_length=200)
+    disabled = BooleanField(required=True)
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/utils.py]: It shouldn't be required to set `disabled`, make it optional but set default value to `False`

File: backend/src/routers/user_auth/utils.py [hunk 35]
```diff
@@ -1,46 +1,51 @@
...
 
 
-class UserInDB(User):  # pylint: disable=R0903
-    hashed_password: str
+def connect_to_db():
+    print("Database login: ")
+    username = input()  # nosec
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/utils.py]: This is weird and insecure.
I guess you don't want to login manually in every run. Please use environmental variables to pass the password:
```python
DB_USERNAME = os.environ.get("DB_USERNAME")
DB_PASSWORD = os.environ.get("DB_PASSWORD")
```
and then just pass them to `connect`.

File: backend/src/routers/user_auth/utils.py [hunk 36]
```diff
@@ -51,11 +56,14 @@ def get_password_hash(password):
...
-        user_dict = db[username]
-        return UserInDB(**user_dict)
-    return None
+def get_user(username: str):
+    try:
+        user = User.objects.get(username=username)
+    except DoesNotExist:
+        return None
+    except MultipleObjectsReturned:
+        logging.error("%s - critical error in database, multiple instances of user.")
```
Comment (pawelrubin) [inline backend/src/routers/user_auth/utils.py]: Is this even possible to have multiple users with the same data?
If so, then this is an issue of your implementation.

File: backend/src/routers/user_auth/utils.py (hunk 24 above)
Comment (mpodgorny) [reply backend/src/routers/user_auth/utils.py]: Changed. thx.

File: frontend/src/components/Auth/Register/Register.js (hunk 22 above)
Comment (mpodgorny) [reply frontend/src/components/Auth/Register/Register.js]: Changed XD.

File: backend/src/main.py (hunk 15 above)
Comment (mpodgorny) [reply backend/src/main.py]: Ofcourse no. Changed.

File: backend/Pipfile (hunk 14 above)
Comment (mpodgorny) [reply backend/Pipfile]: Yup, [for form data.](https://fastapi.tiangolo.com/tutorial/request-files/#form-data)

File: backend/src/routers/user_auth/get_token.py (hunk 13 above)
Comment (mpodgorny) [reply backend/src/routers/user_auth/get_token.py]: changed, thx.

File: backend/src/routers/user_auth/utils.py (hunk 12 above)
Comment (mpodgorny) [reply backend/src/routers/user_auth/utils.py]: Wrong usage of raise HTTPexception, I've changed it.

File: backend/src/routers/user_auth/get_token.py (hunk 11 above)
Comment (mpodgorny) [reply backend/src/routers/user_auth/get_token.py]: Sure, changed it.

**Comment by mpodgorny** [comment]
@wadamek65  Could you take a look on the project structure and review the code, please?

File: .vscode/settings.json [hunk 37]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (wadamek65) [inline .vscode/settings.json]: This file should not be added to repository. Add `.vscode` to `.gitignore`

File: backend/Pipfile [hunk 38]
```diff
@@ -10,13 +10,22 @@ flake8 = "*"
...
 uvicorn = "*"
 email-validator = "*"
+passlib = "*"
+bcrypt = "*"
+datetime = "*"
+pyjwt = "*"
 mongoengine = "*"
 atomicwrites = "*"
+pydantic = "*"
+starlette = "*"
```
Comment (wadamek65) [inline backend/Pipfile]: I think you do not need to install this? I believe this is a direct dependency for `fastapi` so it should already be installed.

**Review by wadamek65 — APPROVED** [review]

File: backend/Pipfile (hunk 38 above)
Comment (mpodgorny) [reply backend/Pipfile]: Youre right, changed it.

**Review by pawelrubin — APPROVED** [review]
